### PR TITLE
Outcome Feedback Feature and Test Fixes

### PR DESCRIPTION
**Pull request by baptisteghostine**
#### Summary

This pull request introduces a new feature that enhances the user experience across all speaking practice modules (Job Interview, Public Speaking, Negotiation). It provides explicit feedback on whether the user would be hired, win a competition, or successfully convince a client based on their performance. Additionally, it tracks successful sessions in Firebase for progress monitoring. The PR also improves code initialization and fixes tests to ensure robustness and reliability.

#### Changes in This PR

1. **Implemented Outcome Feedback Feature**  
   - **Purpose:** Provide users with actionable performance feedback, enhancing realism and educational value.
   - **Actions:**
     - AI analyzes user responses to determine outcomes like being hired, winning, or convincing a client.
     - Displays customized feedback messages with specific reasons for the outcome.
     - Integrated the feature into all speaking practice modules for a consistent experience.
   - **Benefit:** Delivers immediate, context-specific feedback to help users identify strengths and areas for improvement.

2. **Adapted Feature Across All Modules**  
   - **Purpose:** Ensure relevance and meaningful insights for each module.  
   - **Actions:**
     - Added module-specific logic to handle outcome scenarios in `ChatViewModel`.
     - Customized AI prompts for each module to generate appropriate feedback.
     - Implemented a function to parse AI responses and extract decisions.
   - **Benefit:** Engages users with tailored feedback aligned with their practice objectives.

3. **Improved Code Initialization**  
   - **Purpose:** Ensure smooth and error-free application initialization.  
   - **Actions:**
     - Refactored and standardized initialization logic in ViewModels and composables.
     - Added nullability checks and safeguards for critical variables like `practiceContext`.
     - Managed coroutine scopes to prevent race conditions.
   - **Benefit:** Reduces runtime errors and improves performance and maintainability.

4. **Stored Outcome Counts in Firebase**  
   - **Purpose:** Enable progress tracking by storing successful session counts.  
   - **Actions:**
     - Updated `UserProfileRepositoryFirestore` to store and retrieve success counts.
     - Modified `UserStatistics` to include fields for successful interviews, speeches, and negotiations.
     - Added logic to increment success counters based on AI-determined outcomes.
   - **Benefit:** Encourages users by providing tangible metrics of improvement over time.

5. **Fixed and Updated Tests**  
   - **Purpose:** Ensure the test suite reflects the application's current state.  
   - **Actions:**
     - Adjusted `FeedbackScreenTest` for new parameters and outcome logic.
     - Resolved assertion errors in `ChatViewModelTest` and ensured proper initialization.
     - Fixed type mismatches in `UserProfileRepositoryFirestoreTest`, particularly with `Timestamp` objects.
     - Added tests for the outcome feedback feature and its Firebase integration.
   - **Benefit:** Maintains code reliability and prevents regressions.

#### Why This Change Is Important

- **Enhanced Learning Experience:** Clear outcome feedback helps users understand and improve their performance.  
- **Personalized Feedback:** Tailored feedback across modules makes practice sessions more relevant.  
- **Progress Tracking:** Firebase integration motivates users by showing growth over time.  
- **Code Quality and Maintainability:** Improved initialization and testing ensure stability and facilitate future development.

#### Next Steps

1. **Expand Feedback Depth:**  
   - Enhance AI feedback with detailed strengths and weaknesses.

2. **Leaderboard Feature:**  
   - Allow users to compare progress with friends using stored success counts.

3. **User Interface Improvements:**  
   - Highlight success counts in the UI to encourage engagement.

4. **Additional Test Cases:**  
   - Write edge case tests for the outcome determination logic.

#### Future Enhancements

- **Real-Time Feedback:**  
  Provide immediate speech analysis during sessions.

- **Adaptive Difficulty:**  
  Adjust complexity based on user performance history.

- **Gamification Elements:**  
  Add badges and rewards for milestones, leveraging Firebase success counts.

- **Multilingual Support:**  
  Support multiple languages to broaden accessibility.

#### Feedback

Please review these changes and share any suggestions or concerns. This implementation focuses on enhancing user experience, improving code initialization, integrating Firebase for tracking, and ensuring a reliable test suite. Your input is crucial to ensure this feature aligns with user needs and our quality standards.

### CODE REVIEW

**Review by MathisHage — CHANGES_REQUESTED** [review]
#### My review of #162
The changes and new feature seem to work fine (well done for that !), but I have a few remarks to do about the code in itself. Here is a quick summary, but everything is detailed in the comments of the review :

- The scalability of the implementation of the user's achievements. This is in my opinion a major concern, as it could cause problems when increasing the number of modules.
- The documentation. Quite some new public methods are not documented, which would help understanding better their purpose.
- The use of `UserProfileRepository` when testing. The creation of a "test implementation" of this interface is not ideal in my opinion.

I would also add that the test coverage seemingly doesn't reach the recommended 80%.

Please address my comments by directly replying to them. This would allow to keep the discussion organized and focused on resolving each concern one by one.

File: app/src/main/java/com/github/se/orator/model/profile/UserProfile.kt [hunk 1]
```diff
@@ -14,8 +14,20 @@ data class UserProfile(
     val bio: String? = null
 )
 
+enum class SessionType {
+  SPEECH,
+  INTERVIEW,
+  NEGOTIATION
+}
```
Comment (MathisHage) [inline app/src/main/java/com/github/se/orator/model/profile/UserProfile.kt:21]: To avoid using non-scalable `when` statements, I would recommend adding string attributes to each enum, and/or writing some functions in this enum class to retrieve the wanted strings associated to different values of different enums.

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt [hunk 2]
```diff
@@ -30,60 +32,60 @@ class ChatViewModel(
   private val _errorMessage = MutableStateFlow<String?>(null)
   val errorMessage = _errorMessage.asStateFlow()
 
-  private val _collectedAnalysisData = MutableStateFlow<MutableList<AnalysisData>>(mutableListOf())
+  // Change to List<AnalysisData> and update properly
+  private val _collectedAnalysisData = MutableStateFlow<List<AnalysisData>>(emptyList())
```
[the remark this answers is not in the document]
Comment (baptisteghostine) [reply app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:36]: Hi, it was made to ensure that updates to the data are properly propagated through the StateFlow to any observers because when you use a MutableStateFlow, it only emits new values when its value property is updated (for ex when the reference changes). If you have a mutable object (like a MutableList) and you modify it in place ( if you add or remove items), the reference to the list remains the same. As a result, MutableStateFlow does not detect any changes because the value hasn't been reassigned, it's still pointing to the same list instance.

File: app/src/androidTest/java/com/github/se/orator/ui/profile/FakeUserProfileRepository.kt
[the remark this answers is not in the document]
Comment (baptisteghostine) [reply app/src/androidTest/java/com/github/se/orator/ui/profile/FakeUserProfileRepository.kt:1]: You're absolutely right! I updated this and you will see in the next push. Thanks a lot for your comment.

File: app/src/main/java/com/github/se/orator/model/profile/UserProfileViewModel.kt
[the remark this answers is not in the document]
Comment (baptisteghostine) [reply app/src/main/java/com/github/se/orator/model/profile/UserProfileViewModel.kt:1]: There are 2 functions that should have been deleted (I do not use them in my code anymore). And I added comments to the new function.

File: app/src/main/java/com/github/se/orator/model/profile/UserProfile.kt (hunk 1 above)
Comment (baptisteghostine) [reply app/src/main/java/com/github/se/orator/model/profile/UserProfile.kt]: This has been updated in the new push.

File: app/src/main/java/com/github/se/orator/model/profile/UserProfile.kt [hunk 3]
```diff
@@ -14,8 +14,20 @@ data class UserProfile(
...
+
 data class UserStatistics(
     val speechesGiven: Int = 0,
+    val successfulSessions: Int = 0,
+    val successfulSpeeches: Int = 0,
+    val interviewsGiven: Int = 0,
+    val successfulInterviews: Int = 0,
+    val negotiationsGiven: Int = 0,
+    val successfulNegotiations: Int = 0,
```
[the remark this answers is not in the document]
Comment (baptisteghostine) [reply app/src/main/java/com/github/se/orator/model/profile/UserProfile.kt]: This has been updated in the new push.

**Review by MathisHage — APPROVED** [review]
My [latest review](https://github.com/SWENT-TEAM17/OratorAI/pull/162#pullrequestreview-2456445731) was taken into account very seriously, and the code is now absolutely scalable, well documented, and the tests are well written.
The test coverage doesn't reach 80%, but this will be balanced by the fact that the main branch is way over this threshold.

This is excellent work. Approved !

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (MathisHage) [reply app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:36]: Thank you, it makes sense now !
